File: baekjoon/dynamic/1328/Main.java
import java.util.*;

class Main {
    public static final int MOD = 1000000007;
    public static long cache[][][];
    public static int N,L,R;
    
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        L = sc.nextInt();
        R = sc.nextInt();
        
        cache = new long[101][101][101];
        for(int i=0; i<=100; i++) {
            for(int j=0; j<=100; j++) Arrays.fill(cache[i][j], -1);
        }
        System.out.println(solve(N-1,1,1));
    }
    
    public static long solve(int n, int l , int r) {
        if(n == 0) {
            if(l == L && r == R) return 1;
            return 0;
        }
        
        if(cache[n][l][r] != -1) return cache[n][l][r];
        
        cache[n][l][r] = solve(n-1,l+1,r)%MOD;
        cache[n][l][r] += solve(n-1,l,r+1)%MOD;
        cache[n][l][r] += (N-1-n)*solve(n-1,l,r)%MOD;
        return cache[n][l][r] %= MOD;
    }
}

File: baekjoon/dynamic/1328/main.cpp
#include <iostream>
#include <cstdio>
#include <cstring>

using namespace std;
const int MOD = 1000000007;

int N,L,R;
long long solve(int n, int l, int r);
long long cache[101][101][101];
int main()
{
    scanf("%d %d %d", &N, &L, &R);
    memset(cache, -1, sizeof(cache));
    printf("%lld\n", solve(N-1,1,1));
    return 0;
}

long long solve(int n, int l, int r)
{
    if(n == 0) {
        if(l == L && r == R) return 1;
        return 0;
    }
    
    long long& ret = cache[n][l][r];
    if(ret != -1) return ret;
    ret = 0;
    
    ret = solve(n-1, l+1, r)%MOD;
    ret += solve(n-1, l, r+1)%MOD;
    ret += (N-1-n)*solve(n-1, l, r)%MOD;
    return ret%MOD;
}